Cargo.toml

```diff
@@ -45,6 +45,9 @@ matrixmultiply = { version = "0.3.2", default-features = false, features=["cgemm
 serde = { version = "1.0", optional = true, default-features = false, features = ["alloc"] }
 rawpointer = { version = "0.2" }
 
+portable-atomic = { version = "1.6.0", optional = true }
+portable-atomic-util = { version = "0.2.0", features = [ "alloc" ], optional = true }
+
```

Comment:
Can these dependencies be automatically enabled for the relevant platform? I would feel a lot better about this, also lower complexity, if we just managed it all that way with no user visible impact, no feature enablement needed as well.

It's rare that it can be done that way, but here I think the one Arc is a drop-in for the other, and it should work well (?). Or do you see a problem with that?

And yes, it would be good to test this in CI, at least test building it.

Reply:
See https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies

Reply:
The issue is that the `critical-section` feature is recommended to be passed by the library, as mentioned by `portable-atomic`.

> It is very strongly discouraged to enable this feature in libraries that depend on portable-atomic. The recommended approach for libraries is to leave it up to the end user whether or not to enable this feature. (However, it may make sense to enable this feature by default for libraries specific to a platform where it is guaranteed to always be sound, for example in a hardware abstraction layer targeting a single-core chip.)

According to the Arc [docs](https://doc.rust-lang.org/std/sync/struct.Arc.html), we can detect if the target has access to Arc from the configuration flag `target_has_atomic = "ptr"`

> Note: This type is only available on platforms that support atomic loads and stores of pointers, which includes all platforms that support the std crate but not all those which only support [alloc](https://doc.rust-lang.org/1.80.0/alloc/index.html). This may be detected at compile time using #[cfg(target_has_atomic = "ptr")].